### PR TITLE
DEPR: groupby.transform not aligning with input index

### LINKED ISSUE #45648: BUG: groupby transform doesn't respect Series index anymore
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
df = pd.DataFrame({
    'A': ['foo', 'bar', 'foo', 'bar', 'bar', 'foo', 'foo'],
    'C': [2.1, 1.9, 3.6, 4.0, 1.9, 7.8, 2.8]
}).convert_dtypes()

print(df)
     A    C
0  foo  2.1
1  bar  1.9
2  foo  3.6
3  bar  4.0
4  bar  1.9
5  foo  7.8
6  foo  2.8

# sort C within groups defined by A
df.groupby('A')['C'].transform(pd.Series.sort_values)
     C
0  2.1
1  1.9
2  3.6
3  4.0
4  1.9
5  7.8
6  2.8
```

### Issue Description

Suppose I have the following DataFrame

```
df = pd.DataFrame({
    'A': ['foo', 'bar', 'foo', 'bar', 'bar', 'foo', 'foo'],
    'C': [2.1, 1.9, 3.6, 4.0, 1.9, 7.8, 2.8]
}).convert_dtypes()

print(df)
     A    C
0  foo  2.1
1  bar  1.9
2  foo  3.6
3  bar  4.0
4  bar  1.9
5  foo  7.8
6  foo  2.8
```

I used to be able to sort `C`'s values within the groups defined by `A` like this ([proof](https://youtu.be/n681ajrAuVE?t=547)):

```
df.groupby('A')['C'].transform(pd.Series.sort_values)
     C
0  2.1
1  1.9
2  2.8
3  1.9
4  4.0
5  3.6
6  7.8
```

(This was sometime around Pandas 1.0.0)

However pandas 1.4.0 produces the following

```
df.groupby('A')['C'].transform(pd.Series.sort_values)
     C
0  2.1
1  1.9
2  3.6
3  4.0
4  1.9
5  7.8
6  2.8
```

It's as if the `sort_values()` function isn't even being applied.

Note that `df.groupby('A')[['C']].transform(pd.Series.sort_values)` works properly.

```
df.groupby('A')[['C']].transform(pd.Series.sort_values)
     C
0  2.1
1  1.9
2  2.8
3  1.9
4  4.0
5  3.6
6  7.8
```

### Expected Behavior

I would expect `df.groupby('A')['C'].transform(pd.Series.sort_values)` to actually sort C's values within the groups defined by A (as it once did). My expected output in this example would be a Series like this.

```
     C
0  2.1
1  1.9
2  2.8
3  1.9
4  4.0
5  3.6
6  7.8
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : bb1f651536508cdfef8550f93ace7849b00046ee
python           : 3.10.1.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.2.0
Version          : Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 1.4.0
numpy            : 1.22.1
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.1.2
setuptools       : 57.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #45648 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -451,6 +451,13 @@ class providing the base-class of operations.
     The resulting dtype will reflect the return value of the passed ``func``,
     see the examples below.
 
+.. deprecated:: 1.5.0
+    When using ``.transform`` on a grouped DataFrame and the transformation function
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:455]: nit, don't you need a blank line here?

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1196,14 +1196,32 @@ def _transform_general(self, func, *args, **kwargs):
...
+                not emit_alignment_warning
+                and res.ndim == 2
+                and not res.index.equals(group.index)
+            ):
+                emit_alignment_warning = True
+
             res = _wrap_transform_general_frame(self.obj, group, res)
             applied.append(res)
 
+        if emit_alignment_warning:
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1215]: can you add this PR / issue number here as a comment

File: doc/source/user_guide/groupby.rst [hunk 3]
```diff
@@ -776,6 +776,13 @@ as the one being grouped. The transform function must:
 * (Optionally) operates on the entire group chunk. If this is supported, a
   fast path is used starting from the *second* chunk.
 
+.. deprecated:: 1.5.0
+    When using ``.transform`` on a grouped DataFrame and the transformation function
```
Comment (jreback) [inline doc/source/user_guide/groupby.rst:780]: blank line here?

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments

**Comment by rhshadrach** [comment]
Thanks @jreback - changes made

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm merge on green

**Comment by jreback** [comment]
thanks @rhshadrach

**Pull request merged**